Adding more tests with teams loop construct

lthakur007 opened the pull request:
Please review..

seyonglee commented on tests/5.0/teams_loop/test_target_teams_loop_depend.c:
```diff
@@ -0,0 +1,49 @@
+    c[i] = 0;
+  }
+  // Execute on target
+#pragma omp target teams loop map(tofrom: a[0:N]) depend(out: a[0:N])
```
Add `nowait` clauses to this and other target directives, and add `taskwait` directive after the last target construct. 
(Without `nowait` clauses, the generated target task will be an included task; thus, this test cannot check whether the `depend` clause works as intended.)

lthakur007 replied:
added nowait as suggested...

seyonglee commented:
Add brief descriptions to the beginning of each test using the format used in other tests.

seyonglee commented:
GCC 13.1.1 on the Summit system fails on `test_target_teams_loop_allocate.c` with the following error messages:
(the other tests pass with GCC 13.1.1.)

> 25:17: error: 'local' specified in 'allocate' clause but not in an explicit privatization clause

However, this seems to be a compiler bug.

spophale commented on tests/5.0/teams_loop/test_target_teams_loop_allocate.c:
```diff
@@ -0,0 +1,250 @@
+#include <omp.h>
```
Add test description.

spophale commented on tests/5.0/teams_loop/test_target_teams_loop_defaultmap.c:
```diff
@@ -0,0 +1,238 @@
+#include <omp.h>
```
Add test description here, see format of existing tests.

lthakur007 replied:
added description

spophale commented on tests/5.0/teams_loop/test_target_teams_loop_defaultmap.c:
```diff
@@ -0,0 +1,238 @@
+  }
+#pragma omp target teams loop defaultmap(tofrom)
+  for (int i = 0; i < N; i++) {
+    host_data[i] = device_data[i];
```
Should be  host_data[i] += device_data[I]; //to verify if host_data was copied correctly.

lthakur007 replied:
Made changes as suggested...

spophale commented on tests/5.0/teams_loop/test_target_teams_loop_defaultmap.c:
```diff
@@ -0,0 +1,238 @@
+  }
+#pragma omp target teams loop defaultmap(to)
+  for (int i = 0; i < N; i++) {
+    host_data[i] = device_data[i];
```
Suggested change: host_data[i] += device_data[i];

lthakur007 replied:
Made changes as suggested...

spophale commented on tests/5.0/teams_loop/test_target_teams_loop_defaultmap.c:
```diff
@@ -0,0 +1,238 @@
+  }
+  // checking results
+  for (int i = 0; i < N; i++) {
+    OMPVV_TEST_AND_SET(errors, host_data[i] != 0);
```
On a shared memory system or host fallback host_data will be updated (which is correct). This will lead to incorrect error reporting.

Better to do a map(from: host_data) on the target and check for updated host_data.

lthakur007 replied:
The intent of using "defaultmap(to)" is to ensure that host side memory(process's stack memory) "host_data" is not updated. Hence the check "host_data[i] != 0" at line 71.
Because host_data is host memory allocated in stack, the expectation is that it doesn't come under "shared memory" aspects... Please let us know if your opinion differs...

seyonglee replied:
> Because host_data is host memory allocated in stack, the expectation is that it doesn't come under "shared memory" aspects... Please let us know if your opinion differs…

I don't understand the above comment; `host_data` on the host will be updated if on a shared memory system regardless of whether `host_data` is in the stack memory or global memory. 
I think that a better way to handle this issue is to print a warning message if this test runs on the shared memory system or host fallback occurs.

lthakur007 replied:
added map(from: host_data) as suggested...

seyonglee replied:
Change `map(from: host_data)` to `map(tofrom: host_data)` since `host_data` is read first in the target region.
Change  `OMPVV_TEST_AND_SET(errors, host_data[i] != 0);` to `OMPVV_TEST_AND_SET(errors, host_data[i] != i);`

seyonglee replied:
One remaining issue is that this test may not necessarily check the behavior of the `defaultmap(to)` clause.

spophale replied:
@lthakur007, you can potentially add a constant based on omp_is_initial_device() to the host_data to differentiate between host_fallback vs. target execution.

spophale commented on tests/5.0/teams_loop/test_target_teams_loop_defaultmap.c:
```diff
@@ -0,0 +1,238 @@
+  }
+#pragma omp target teams loop defaultmap(default)
+  for (int i = 0; i < N; i++) {
+    host_data[i] = device_data[i];
```
Suggested change: host_data[i] += device_data[i];

lthakur007 replied:
Made changes as suggested...

spophale commented on tests/5.0/teams_loop/test_target_teams_loop_defaultmap.c:
```diff
@@ -0,0 +1,238 @@
+  }
+#pragma omp target teams loop defaultmap(tofrom:aggregate)
+  for (int i = 0; i < N; i++) {
+    host_data[i] = device_data[i];
```
Suggested change: host_data[i] += device_data[i];

lthakur007 replied:
Made changes as suggested...

spophale commented on tests/5.0/teams_loop/test_target_teams_loop_defaultmap.c:
```diff
@@ -0,0 +1,238 @@
+  }
+  // checking results to verify host_data is unchanged
+  for (int i = 0; i < N; i++) {
+    OMPVV_TEST_AND_SET(errors, host_data[i] != 0);
```
Same issue as mentioned before for host fallback.

lthakur007 replied:
added map(from: host_data) as suggested...

spophale commented on tests/5.0/teams_loop/test_target_teams_loop_defaultmap.c:
```diff
@@ -0,0 +1,238 @@
+  int *ptr = 0;
+#pragma omp target teams loop defaultmap(tofrom:pointer)
+  for (int i = 0; i < N; i++) {
+    ptr = (int*)0xefefefef;
```
Multiple threads could be assigning to the same pointer variable. 
Either have an atomic write or an array of ptrs.

lthakur007 replied:
Made changes as suggested...

tob2 commented:
> GCC 13.1.1 on the Summit system fails on `test_target_teams_loop_allocate.c` with the following error messages: (the other tests pass with GCC 13.1.1.)
> 
> > 25:17: error: 'local' specified in 'allocate' clause but not in an explicit privatization clause
> 
> However, this seems to be a compiler bug.

Actually, it is a code bug:

```C
#pragma omp target teams loop ... allocate(omp_default_mem_alloc: local) private(local)
```

Namely:
* the **private** clause is accepted by all of them: **loop**, **teams** and **target**, but "Clauses on Combined and Composite Constructs" section states: "The effect of the **private** clause is as if it is applied only to the innermost leaf construct that permits it." — Thus, **private** applies to only to **loop**.
* the **allocate** clause is accepted only by **teams** and **target** but not by **loop**. The "Clauses on Combined and Composite Constructs" section states: "The effect of the allocate clause is as if it is applied to all leaf constructs that permit the clause and to which a data-sharing attribute clause that may create a private copy of the same list item is applied." → In principle, **teams** and **target** would qualify, but neither of them has  a data-sharing attribute (as the 'private' is already consumed by 'loops')
* Additionally, the following is required — but not fulfilled **allocate**: "Restrictions to clauses on combined and composite constructs are as follows:
   * A clause that appears on a combined or composite construct must apply to at least one of the leaf constructs per the rules defined in this section."

lthakur007 commented:
Hi @seyonglee and @seyonglee ,
Could you please share your opinion on the comment made by @tob2 .
Please advise if we should remove the test file or any modification would allow allocate() clause still applicable.

tob2 commented on tests/5.0/teams_loop/test_target_teams_loop_allocate.c:
```diff
@@ -0,0 +1,250 @@
+    c[i] = 0;
+  }
+  // Execute on target
+#pragma omp target teams loop map(to: a[0:N], b[0:N]) map(from: c[0:N])\
+        uses_allocators(omp_default_mem_alloc)\
+        allocate(omp_default_mem_alloc: local) private(local)
```
> Please advise if we should remove the test file or any modification would allow allocate() clause still applicable.

A bit pointless, but you could add a 'firstprivate' clause (which then applies to target and teams, but not to 'loop') – and keep the 'private' for 'loop'.

But somewhat more sensible would be to split 'target teams loop' as follows – such that the private applies to 'teams' – That's still not very useful as the variable is twice privatized - once on 'teams' level and once on 'loop' level, but still.
```suggestion
#pragma omp target teams map(to: a[0:N], b[0:N]) map(from: c[0:N]) \
        uses_allocators(omp_cgroup_mem_alloc) \
        allocate(omp_low_lat_mem_alloc: local) private(local)
  #pragma omp loop private(local)
```

How about something like the following? Keeping the private but adding another firstprivatized variable, which is now put into "const" memory to make it more interesting than "default":
```C++
int magic_value = 42;
#pragma omp target teams loop map(to: a[0:N], b[0:N]) map(from: c[0:N]) \
        uses_allocators(omp_default_mem_alloc) \
        allocate(omp_const_mem_alloc: magic_value) firstprivate(magic_value) \
        private(local)
  for (int i = 0; i < N; i++) {
    local = a[i] + b[i] + magic_value;
    c[i] = local;
  }
```

seyonglee replied:
For me, the last option (use `first private(magic_value)`) seems to be the best.

lthakur007 commented:
addressed review comments.. please approve...

seyonglee commented on tests/5.0/teams_loop/test_target_teams_loop_depend.c:
```diff
@@ -0,0 +1,49 @@
+#include <omp.h>
```
Add a test description here.

seyonglee commented on tests/5.0/teams_loop/test_target_teams_loop_allocate.c:
```diff
@@ -0,0 +1,259 @@
+#pragma omp target teams map(to: a[0:N], b[0:N]) map(from: c[0:N]) \
+        uses_allocators(omp_cgroup_mem_alloc) \
+        allocate(omp_low_lat_mem_alloc: local) private(local)
+  #pragma omp loop private(local)
```
The purpose of this test is to test the `teams loop` combined construct, but this sub-test uses separate `teams` construct and `loop` construct, defeating the purpose of this test.

lthakur007 replied:
Replaced "teams" with  "teams loop" as suggested...

seyonglee commented on tests/5.0/teams_loop/test_target_teams_loop_allocate.c:
```diff
@@ -0,0 +1,259 @@
+  // Execute on target
+#pragma omp target teams loop map(to: a[0:N], b[0:N]) map(from: c[0:N])\
+        uses_allocators(omp_large_cap_mem_alloc)\
+        allocate(omp_large_cap_mem_alloc: local) private(local)
```
The error @tob2 mentioned still exists here (`error: 'local' specified in 'allocate' clause but not in an explicit privatization clause`); one possible solution is to use `firstprivate(local)` instead of `private(local)`, but using the `firstprivate` clause will incur unnecessary initialization of the firstprivate variable, `local`.

lthakur007 replied:
Replaced "private(local)" with  "firstprivate(local)" as suggested...

seyonglee commented on tests/5.0/teams_loop/test_target_teams_loop_allocate.c:
```diff
@@ -0,0 +1,259 @@
+  // Execute on target
+#pragma omp target teams loop map(to: a[0:N], b[0:N]) map(from: c[0:N])\
+        uses_allocators(omp_high_bw_mem_alloc)\
+        allocate(omp_high_bw_mem_alloc: local) private(local)
```
The error @tob2 mentioned still exists here (`error: 'local' specified in 'allocate' clause but not in an explicit privatization clause`); one possible solution is to use `firstprivate(local)` instead of `private(local)`, but using the `firstprivate` clause will incur unnecessary initialization of the firstprivate variable, `local`.

lthakur007 replied:
Replaced "private(local)" with  "firstprivate(local)" as suggested...

seyonglee commented on tests/5.0/teams_loop/test_target_teams_loop_allocate.c:
```diff
@@ -0,0 +1,259 @@
+  // Execute on target
+#pragma omp target teams loop map(to: a[0:N], b[0:N]) map(from: c[0:N])\
+        uses_allocators(omp_low_lat_mem_alloc)\
+        allocate(omp_low_lat_mem_alloc: local) private(local)
```
The error @tob2 mentioned still exists here (`error: 'local' specified in 'allocate' clause but not in an explicit privatization clause`); one possible solution is to use `firstprivate(local)` instead of `private(local)`, but using the `firstprivate` clause will incur unnecessary initialization of the firstprivate variable, `local`.

lthakur007 replied:
Replaced "private(local)" with  "firstprivate(local)" as suggested...

seyonglee commented on tests/5.0/teams_loop/test_target_teams_loop_allocate.c:
```diff
@@ -0,0 +1,259 @@
+  // Execute on target
+#pragma omp target teams loop map(to: a[0:N], b[0:N]) map(from: c[0:N])\
+        uses_allocators(omp_pteam_mem_alloc)\
+        allocate(omp_pteam_mem_alloc: local) private(local)
```
The error @tob2 mentioned still exists here (`error: 'local' specified in 'allocate' clause but not in an explicit privatization clause`); one possible solution is to use `firstprivate(local)` instead of `private(local)`, but using the `firstprivate` clause will incur unnecessary initialization of the firstprivate variable, `local`.

lthakur007 replied:
Replaced "private(local)" with  "firstprivate(local)" as suggested...

lthakur007 commented on tests/5.0/teams_loop/test_target_teams_loop_allocate.c:
```diff
@@ -0,0 +1,259 @@
+  // Execute on target
+#pragma omp target teams loop map(to: a[0:N], b[0:N]) map(from: c[0:N])\
+        uses_allocators(omp_cgroup_mem_alloc)\
+        allocate(omp_cgroup_mem_alloc: local) private(local)
```
Replaced "private(local)" with  "firstprivate(local)" as suggested...

lthakur007 commented on tests/5.0/teams_loop/test_target_teams_loop_allocate.c:
```diff
@@ -0,0 +1,259 @@
+  // Execute on target
+#pragma omp target teams loop map(to: a[0:N], b[0:N]) map(from: c[0:N])\
+        uses_allocators(omp_default_mem_alloc)\
+        allocate(omp_default_mem_alloc: local) private(local)
```
Replaced "private(local)" with  "firstprivate(local)" as suggested...

lthakur007 commented:
Hi @seyonglee and @spophale ,
I have addressed the review comments. Please consider to scrutinize the fixes...
Fixes are working fine on AMD platforms....

seyonglee commented on tests/5.0/teams_loop/test_target_teams_loop_allocate.c:
```diff
@@ -0,0 +1,259 @@
+    c[i] = 0;
+  }
+  // Execute on target
+#pragma omp target teams loops map(to: a[0:N], b[0:N]) map(from: c[0:N]) \
```
Change `loops` to `loop`

lthakur007 replied:
Observed that its already changed...

seyonglee commented on tests/5.0/teams_loop/test_target_teams_loop_defaultmap.c:
```diff
@@ -0,0 +1,250 @@
+    host_data[i] = 0;
+  }
+#pragma omp target teams loop defaultmap(alloc) map(to: device_data[0:N])\
+         map(from: host_data)
```
Delete `map(from: host_data)`

spophale commented on tests/5.0/teams_loop/test_target_teams_loop_defaultmap.c:
```diff
@@ -0,0 +1,249 @@
+    device_data[i] = i;
+    host_data[i] = 0;
+  }
+#pragma omp target teams loop defaultmap(to) map(from: host_data)
```
needs to be:  map(tofrom: host_data)

spophale commented on tests/5.0/teams_loop/test_target_teams_loop_defaultmap.c:
```diff
@@ -0,0 +1,249 @@
+  }
+  // checking results
+  for (int i = 0; i < N; i++) {
+    OMPVV_TEST_AND_SET(errors, host_data[i] != 0);
```
Needs to check host_data[i] == i

spophale commented:
@lthakur007 , generally arrays are not the best data types to check if defaultmaps were correctly applied as they have an implicit data mapping of toform on target directive.

spophale commented on tests/5.0/teams_loop/test_target_teams_loop_defaultmap.c:
```diff
@@ -0,0 +1,249 @@
+}
+
+int main() {
+  int errors = 0;
```
Add:
int is_offloading;
OMPVV_TEST_AND_SET_OFFLOADING(is_offloading);


This will ensure that the correct message is printed. Like "Test passed on device .."

spophale commented:
Add for all tests in main:
int is_offloading;
OMPVV_TEST_AND_SET_OFFLOADING(is_offloading);

This will ensure that the correct message is printed. Like "Test passed on device .."

lthakur007 commented on tests/5.0/teams_loop/test_target_teams_loop_allocate.c:
```diff
@@ -0,0 +1,259 @@
+  // Execute on target
+#pragma omp target teams loops map(to: a[0:N], b[0:N]) map(from: c[0:N]) \
+        uses_allocators(omp_cgroup_mem_alloc) \
+        allocate(omp_low_lat_mem_alloc: local) private(local)
```
Observed that its already changed...

lthakur007 commented:
Hi @seyonglee and @spophale ,
I have addressed review comments... please consider the fixes....

spophale commented on tests/5.0/teams_loop/test_target_teams_loop_defaultmap.c:
```diff
@@ -0,0 +1,249 @@
+  }
+  // checking results
+  for (int i = 0; i < N; i++) {
+    OMPVV_TEST_AND_SET(errors, host_data[i] != 0);
```
This test will fail when host fallback occurs OR machine has a unified address space.

spophale commented on tests/5.0/teams_loop/test_target_teams_loop_defaultmap.c:
```diff
@@ -0,0 +1,249 @@
+  int x = 0;
+#pragma omp target teams loop defaultmap(tofrom:scalar)
+  for (int i = 0; i < N; i++) {
+    x = N;
```
"x" is shared across the teams and there is a race to update it. It may not be apparent as all threads are setting it ti N, but this leads to a non-conformant OpenMP program.
 To check the 'tofrom' consider: 
#pragma omp target teams loop defaultmap(tofrom:scalar) map(from: array) 
for (int i = 0; i < N; i++) {
#pragma omp atomic
    x += i; //this checks if the 'to' for x is working correctly
array[i] = x + 2;
}
//Then check array and x values on host

spophale commented on tests/5.0/teams_loop/test_target_teams_loop_defaultmap.c:
```diff
@@ -0,0 +1,249 @@
+    device_data[i] = i;
+    host_data[i] = 0;
+  }
+#pragma omp target teams loop defaultmap(tofrom:aggregate) map(from: host_data)
```
defaultmap will not map host_data "to" the device since there is a explicit time.

spophale replied:
```suggestion
#pragma omp target teams loop defaultmap(tofrom:aggregate) map(tofrom: host_data)
```

spophale commented:
@lthakur007 please review our comments at your convenience.

fel-cab commented on tests/5.0/teams_loop/test_target_teams_loop_allocate.c:
```diff
@@ -0,0 +1,261 @@
+  }
+  // Execute on target
+#pragma omp target teams loop map(to: a[0:N], b[0:N]) map(from: c[0:N]) \
+        uses_allocators(omp_cgroup_mem_alloc) \
+        allocate(omp_low_lat_mem_alloc: local) firstprivate(local)
```
```suggestion
        uses_allocators(omp_low_lat_mem_alloc) \
        allocate(omp_low_lat_mem_alloc: local) firstprivate(local)
```
Is this correct?
Otherwise the compilation error is geenerated!
```
error: allocator must be specified in the 'uses_allocators' clause
        allocate(omp_low_lat_mem_alloc: local) firstprivate(local)
                 ^~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

lthakur007 commented:
Hi @fel-cab ,
I have committed the changes you suggested. Please consider the PR for further processing...
